6942632: Hotspot should be able to use more than 64 logical processors on Windows

swesonga opened the pull request:
Windows processor groups support a maximum of 64 logical processors per processor group. OpenJDK currently uses the GetSystemInfo API to get the number of processors on the system. GetSystemInfo only returns the number of processors in the current processor group, which is at most 64. Therefore, OpenJDK needs to be updated to support the use of all available processors across all Windows processor groups. However, this should only be done when running on Windows 11 or Windows Server 2022 or later Windows operating systems.

Starting with Windows 11 and Windows Server 2022, applications are no longer constrained by default to a single processor group. Instead, processes and their threads have processor affinities that by default span all processors in the system, across multiple groups on machines with more than 64 processors. This is in contrast to previous operating system versions where the processor group affinity APIs needed to be used to assign each thread a specific processor group. See https://learn.microsoft.com/en-us/windows/win32/procthread/processor-groups#behavior-starting-with-windows-11-and-windows-server-2022

The key changes required are:

1. Detecting Windows 11 or Windows Server 2022 or later.
1. Using the The GetLogicalProcessorInformationEx function to correctly retrieve the total number of processors available on the system. This function was introduced in Windows 7 so its availability needs to be explicitly determined.
1. Updating the calculation of the number of active processors.
    1. Processor affinities can be set in a Windows job object and if the JVM is launched in a job object, these affinities restrict the processors that the JVM can use (regardless of how many processors the system has).
    1. The JVM can also be launched from the command prompt by using the "start" command to set processor affinities instead of using a job object. This command sets the process group affinity of the new process to a single processor group (since the affinity mask is a 64-bit value). The affinity mask passed to the "start" command explicitly sets the number of available processors for the JVM and is necessarily at most 64, even if the machine has more than 64 processors.
    1. If no affinity restrictions are present (i.e. there is no job object in use and an affinity setting command was not used), then all cores on the system will be available to the JVM.

### Test Scenarios
This simple Java program can be used to test these changes:

```
public class GetAvailableProcessors {
    public static void main(String[] args) {
        System.out.println(Runtime.getRuntime().availableProcessors());
    }
}
```

To use all processors in the default processor group (and no more), either of these 2 equivalent commands can be used:

```
%JAVA_HOME%/bin/java -Xlog:os=debug:file=os.txt::filecount=0 <JavaProgram> <JavaProgramArgs>
%JAVA_HOME%/bin/java -XX:-UseAllWindowsProcessorGroups -Xlog:os=debug:file=os.txt::filecount=0 <JavaProgram> <JavaProgramArgs>
```

To use all processors on the Windows 11/Server 2022 machine, run the next command. On older Windows versions, all processors in the default processor group (and no more) will be used:

```
%JAVA_HOME%/bin/java -XX:+UseAllWindowsProcessorGroups -Xlog:os=debug:file=os.txt::filecount=0 <JavaProgram> <JavaProgramArgs>
```

To restrict processor affinity when using the primary processor group only, use the start command with the /affinity argument. In this case, neither the OS version nor the product flag will affect the number of processors used. There will be 10 active processors in the first command below and 9 active processors with the second command.

```
start /affinity abcd %JAVA_HOME%/bin/java -XX:-UseAllWindowsProcessorGroups -Xlog:os=debug:file=os.txt::filecount=0 <JavaProgram> <JavaProgramArgs>
start /affinity ef12 %JAVA_HOME%/bin/java -XX:+UseAllWindowsProcessorGroups -Xlog:os=debug:file=os.txt::filecount=0 <JavaProgram> <JavaProgramArgs>
```

There is no standard Windows tool for launching processes in a job object. However, a custom tool, say CreateJobObject.exe, can be created to create a process and assign it to a job object before starting execution of the process. In these examples, the CreateJobObject.exe tool can be provided with a list of processor affinity masks for each processor group to be active. Consider the example below. On older Windows versions, all processors in the default processor group will be active (regardless of the product flag value). On Windows 11/Server 2022, exactly 5 processors will be active (the processors in the default processor group that are allowed by the job object).

```
CreateJobObject.exe "%JAVA_HOME%/bin/java -XX:-UseAllWindowsProcessorGroups -Xlog:os=debug:file=os.txt::filecount=0 <JavaProgram> <JavaProgramArgs>" 1234 5678
```

If the UseAllWindowsProcessorGroups is enabled, then all processors in the job object will be counted as active processors. For example, 9+7=16 processors will be active for a job object created using this command:

```
CreateJobObject.exe "%JAVA_HOME%/bin/java -XX:+UseAllWindowsProcessorGroups -Xlog:os=debug:file=os.txt::filecount=0 -Xlog:os=debug:file=os.txt::filecount=0 <JavaProgram> <JavaProgramArgs>" ef12 12f4
```

### References:
https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getsysteminfo
https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getlogicalprocessorinformationex
https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/start

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8326890](https://bugs.openjdk.org/browse/JDK-8326890) to be approved

### Issues
 * [JDK-6942632](https://bugs.openjdk.org/browse/JDK-6942632): Hotspot should be able to use more than 64 logical processors on Windows (**Enhancement** - P3)
 * [JDK-8326890](https://bugs.openjdk.org/browse/JDK-8326890): Hotspot should be able to use more than 64 logical processors on Windows (**CSR**)


### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17576/head:pull/17576` \
`$ git checkout pull/17576`

Update a local copy of the PR: \
`$ git checkout pull/17576` \
`$ git pull https://git.openjdk.org/jdk.git pull/17576/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17576`

View PR using the GUI difftool: \
`$ git pr show -t 17576`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17576.diff">https://git.openjdk.org/jdk/pull/17576.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17576#issuecomment-1910824736)

dholmes-ora commented:
Hi @swesonga , thanks for this contribution. I am taking a look at it but these are API's that I am not familiar with (and I'm not sure we have any Windows experts who can review, other than Microsoft folk themselves). My main concern would be compatibility - ensuring earlier Windows versions continue to behave the same, but also the impact on applications already running on Windows 11 or Server 2022 which may suddenly get an increase in the number of available processors. We might need a product flag to opt-in to this new behaviour, or at least a diagnostic flag to allow users to opt-out.

swesonga commented:
> Hi @swesonga , thanks for this contribution. I am taking a look at it but these are API's that I am not familiar with (and I'm not sure we have any Windows experts who can review, other than Microsoft folk themselves). My main concern would be compatibility - ensuring earlier Windows versions continue to behave the same, but also the impact on applications already running on Windows 11 or Server 2022 which may suddenly get an increase in the number of available processors. We might need a product flag to opt-in to this new behaviour, or at least a diagnostic flag to allow users to opt-out.

Thanks @dholmes-ora. I will add a product flag to opt-in to this new functionality.

dholmes-ora commented:
A question has also been raised about the NUMA API's and whether there is a plan to update those to work with all the processors?

swesonga commented:
> A question has also been raised about the NUMA API's and whether there is a plan to update those to work with all the processors?

Support for NUMA APIs will be addressed in a separate workstream as part of [JDK-8244065: Implement NUMA support on Windows](https://bugs.openjdk.org/browse/JDK-8244065)

swesonga commented:
Hi @dholmes-ora, what is the expected approach for adding flags for a feature like this? https://wiki.openjdk.org/display/HotSpot/Hotspot+Command-line+Flags%3A+Kinds%2C+Lifecycle+and+the+CSR+Process states that product flags are considered exported interfaces that require a CSR request when they are added. Does the flag need to start out as an experimental flag or is it okay to just add it as a product flag for opting in and then going through the CSR request immediately?

dholmes-ora commented:
@swesonga it is fine to start out as a product flag as per the PR.

/csr needed

jdksjolen commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3978,14 +3985,184 @@ bool   os::win32::_is_windows_server         = false;
+
+      job_object_information = os::malloc(job_object_information_length, mtInternal);
+      if (job_object_information != nullptr) {
+          if (QueryInformationJobObject(nullptr, JobObjectGroupInformationEx, job_object_information, job_object_information_length, &job_object_information_length)) {
```
Implicit cast.

jdksjolen commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3978,14 +3985,184 @@ bool   os::win32::_is_windows_server         = false;
+          if (QueryInformationJobObject(nullptr, JobObjectGroupInformationEx, job_object_information, job_object_information_length, &job_object_information_length)) {
+            DWORD groups_found = job_object_information_length / sizeof(GROUP_AFFINITY);
+            if (groups_found != group_count) {
+              warning("Unexpected processor group count->%ld. Expected %ld processor groups.", groups_found, group_count);
```
Why `->`? Perhaps ": " or " of " is better

jdksjolen reviewed (commented):
Hi,

Please avoid implicit casts from int to bool. I noted two examples, but there are more in the code. I have a preference for avoiding unnecessary indentation, and I noted one place where this can be done, but consider that a nit and not required :-).

Other than that, the code seems correct and the idea itself sound.

swesonga commented:
> Hi,
> 
> Please avoid implicit casts from int to bool. I noted two examples, but there are more in the code. I have a preference for avoiding unnecessary indentation, and I noted one place where this can be done, but consider that a nit and not required :-).
> 
> Other than that, the code seems correct and the idea itself sound.

I have removed the implicit casts in the new code I introduced. Please let me know if I missed any - I did not remove implicit casts from any of the previously existing code.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3978,14 +3985,192 @@ bool   os::win32::_is_windows_server         = false;
+            assert(false, "Must find at least 1 logical processor");
+          }
+        } else {
+          warning("QueryInformationJobObject() failed: GetLastError->%ld.", GetLastError());
```
Again use `os::lasterror` please.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3978,14 +3985,192 @@ bool   os::win32::_is_windows_server         = false;
+  PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX system_logical_processor_info = nullptr;
+  DWORD returned_length = 0;
+
+  // https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getlogicalprocessorinformationex
```
There is no need to given URLs for documentation unless there is something very specific about that documentation that is worth linking. We can assume the developer can lookup API reference documents. And URLs get stale.

dholmes-ora reviewed (commented):
Seems reasonable in principle but there is a lot of detail here that I'm not familiar with. I will at least take this for a spin in our CI with the flag on and off.

swesonga commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -867,21 +868,60 @@ int os::active_processor_count() {
-      bitcount++;
-    }
-    return bitcount;
+  if (GetProcessAffinityMask(GetCurrentProcess(), &lpProcessAffinityMask, &lpSystemAffinityMask)) {
```
> I'm struggling to see how we can expose more than 64 processors when our bit-vector is only 64 bits. ??

This code block does not expose over 64 processors. It only ensures that a request for less than 64 processors is respected when the user launches the process using the start command (which accepts a 64-bit mask) https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/start

More than 65 cores are exposed by the `logical_processors = processor_count();` assignment a few lines later.

dholmes-ora replied:
Sorry I'm still not clear here. From my reading `GetProcessAffinityMask` is either going to return a zero mask or a non-zero mask. In the non-zero case, the count must be <=64. So even if we intend to use > 64 processors, this mask reading will always limit us to a value < 64. And the zero mask is only for one special case. So regardless of the use of the "start" command, it seems to me this will cause us to bail out at line 903. ??

swesonga replied:
> Sorry I'm still not clear here. From my reading `GetProcessAffinityMask` is either going to return a zero mask or a non-zero mask. In the non-zero case, the count must be <=64. So even if we intend to use > 64 processors, this mask reading will always limit us to a value < 64. And the zero mask is only for one special case. So regardless of the use of the "start" command, it seems to me this will cause us to bail out at line 903. ??

If there are 64 bits set in the process affinity mask, there is no way to tell whether or not the start command was used to set this mask. This implementation treats this case as equivalent to the user not restricting processor affinities at all. We don't bail out at line 903 because in this case, logical_processors == si.dwNumberOfProcessors and all 65+ processors from line 964 will therefore be usable.

dholmes-ora replied:
Sorry to belabour this but IIUC the only time we will use > 64 processors is if the affinity masks is all ones - right? But we cannot distinguish between the cases of only wanting to use 64 processors and wanting to use > 64 processors. And if we didn't use the start command and have not otherwise set the affinity mask, then that mask will be all ones.

swesonga commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -867,21 +868,60 @@ int os::active_processor_count() {
+    // Number of active processors is number of bits in process affinity mask
+    logical_processors = population_count(lpProcessAffinityMask);
+
+    if (logical_processors < si.dwNumberOfProcessors) {
```
> The documentation states:
> 
> > On a system with more than 64 processors, if the threads of the calling process are in a single [processor group](https://learn.microsoft.com/en-us/windows/desktop/ProcThread/processor-groups), the function sets the variables pointed to by lpProcessAffinityMask and lpSystemAffinityMask to the process affinity mask and the processor mask of active logical processors for that group. If the calling process contains threads in multiple groups, the function returns zero for both affinity masks.
> 
> So it seems that we need to do something special to handle zero here.

Good point. I have added a check to ensure that zero is handled by the final return statement of this method.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -867,21 +868,64 @@ int os::active_processor_count() {
-    return proc_count;
+    char buf[512];
+    size_t buf_len = os::lasterror(buf, sizeof(buf));
+    if (buf_len != 0) {
+      warning(buf);
+    } else {
+      warning("Attempt to get the process affinity mask failed.");
+    }
```
What I meant to do here is this:
```
 warning("Attempt to get the process affinity mask failed: %s", buf_len != 0 ? buf : "<unknown error>");
```

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -1911,6 +1912,12 @@ void os::pd_print_cpu_info(outputStream* st, char* buf, size_t buflen) {
+    // at most 64. The GetLogicalProcessorInformation function is used to compute the total number
+    // of processors. However, it requires memory to be allocated for the processor information buffer.
+    // Since this method is used in paths where memory allocation should not be done (i.e. after a crash),
+    // only the number of processors in the current group will be returned.
```
Okay but then at line 1950 I think we should change:
```
"Processor Information for all %d processors :"
```
to
```
"Processor Information for the first %d processors :"
```

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3978,14 +3985,199 @@ bool   os::win32::_is_windows_server         = false;
+  len = MAX_PATH - (UINT)strlen("\\kernel32.dll") - 1;
+  ret = GetSystemDirectory(kernel32_path, len);
+  if (ret == 0 || ret > len) {
+    warning("GetSystemDirectory() failed: GetLastError->%ld.", GetLastError());
```
Sorry if it was unclear but all warnings should ideally use `os::last_error()` to report a user-friendly error message.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -867,21 +868,60 @@ int os::active_processor_count() {
   }
 
+  bool schedules_all_processor_groups = false;
+  if (UseAllWindowsProcessorGroups) {
+    // Starting with Windows 11 and Windows Server 2022, the OS has changed to
+    // make processes and their threads span all processors in the system,
+    // across all processor groups, by default. Therefore, we will allow all
+    // processors to be active processors on these operating systems. However,
+    // job objects can be used to restrict processor affinity across the
+    // processor groups. In this case, the number of active processors must be
+    // obtained from the processor affinity in the job object.
+    schedules_all_processor_groups = win32::is_windows_11_or_greater() || win32::is_windows_server_2022_or_greater();
+    if (schedules_all_processor_groups) {
+      DWORD processors_in_job_object = win32::active_processors_in_job_object();
```
Shouldn't the check for a job object be unconditional? I don't understand what the connection is to either `UseAllWindowsProcessorGroups` or the Windows version here. ??

swesonga replied:
> Shouldn't the check for a job object be unconditional? I don't understand what the connection is to either `UseAllWindowsProcessorGroups` or the Windows version here. ??

A job object can enable multiple Windows processor groups. I think running with -XX:-UseAllWindowsProcessorGroups should override the job object enabling processors in >1 processor group (closer to behavior that would be seen in older OpenJDK versions). The approach I've taken adds a warning when this happens. Other than that, I've updated the implementation to respect the number of processors enabled by the job object. On older Windows versions, this includes displaying a warning that only 1 processor group is supported. Please let me know what you think of this approach.

dholmes-ora commented:
@swesonga  you have not progressed the CSR request for this issue.

dholmes-ora commented on src/hotspot/os/windows/os_windows.hpp:
```diff
@@ -56,6 +63,38 @@ class os::win32 {
+  static bool   is_windows_11_or_greater();
+  static bool   is_windows_server_2022_or_greater();
+  static DWORD  system_logical_processor_count();
+  static int windows_major_version() {
+    initialize_windows_version();
+    return _major_version;
+  }
+  static int windows_minor_version() {
+    initialize_windows_version();
+    return _minor_version;
+  }
+  static int windows_build_number() {
+    initialize_windows_version();
+    return _build_number;
+  }
+  static int windows_build_minor() {
+    initialize_windows_version();
+    return _build_minor;
+  }
```
There should only be one path that initializes the version info and we should be very clear what that one path is. Can these methods be called before the call to `os::init` that calls `initialize_system_info` that calls `initialize_windows_version`? If so we should fix that. Otherwise these methods should assert the version is initialized.

Then `initialize_windows_version` should assert that it is not already initialized.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3978,14 +4002,214 @@ bool   os::win32::_is_windows_server         = false;
+    assert(false, "Unexpected QueryInformationJobObject error code");
+    return 0;
+  }
+
+  return processors;
```
Before returning please add a `log_debug(os)` statement indicating that there is an active job object and the number of processors it defines.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3978,14 +4002,214 @@ bool   os::win32::_is_windows_server         = false;
+      warning("Attempt to determine logical processor count from GetLogicalProcessorInformationEx() failed: %s", buf_len != 0 ? buf : "<unknown error>");
+    }
+  }
+
+  return logical_processors;
```
Before returning please add a `log_debug(os)` statement indicating the number of logical processors discovered.

swesonga replied:
This message will not appear in the log if the logging statement is placed here because the [log configuration is initialized](https://github.com/openjdk/jdk/blob/8b9bf758801400e4491326cd4c90fc117b9d97e1/src/hotspot/share/runtime/threads.cpp#L449) after [initializing the os module](https://github.com/openjdk/jdk/blob/8b9bf758801400e4491326cd4c90fc117b9d97e1/src/hotspot/share/runtime/threads.cpp#L431). I have added logging code that will be executed only the first time the active processor count is retrieved.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3978,14 +4002,214 @@ bool   os::win32::_is_windows_server         = false;
+
+  DWORD processors = 0;
+  bool schedules_all_processor_groups = win32::is_windows_11_or_greater() || win32::is_windows_server_2022_or_greater();
+  if (schedules_all_processor_groups) {
```
Please add a `log_debug(os)` statement indicating that this version does schedule all processor groups.

swesonga replied:
> Please add a `log_debug(os)` statement indicating that this version does schedule all processor groups.

This message will not appear in the log if the logging statement is placed here because the [log configuration is initialized](https://github.com/openjdk/jdk/blob/8b9bf758801400e4491326cd4c90fc117b9d97e1/src/hotspot/share/runtime/threads.cpp#L449) after [initializing the os module](https://github.com/openjdk/jdk/blob/8b9bf758801400e4491326cd4c90fc117b9d97e1/src/hotspot/share/runtime/threads.cpp#L431). I have added logging code that will be executed only the first time the active processor count is retrieved.

dholmes-ora reviewed (changes requested):
Sorry for the delay in getting back to this. There is a lot of API surface that I am not familiar with that I am trying to get my head around. The overall code structure and approach is looking very good - though I have one issue with the initialization process.

One thing that is missing though is a way to effectively test this given all the possible conditions that might be present:
- a job object exists, or not
- processor groups exist, or not
- affinity has been set, or not
I recognise it may be very difficult to create either a regression test or a gtest for this, but I would like to see some instructions on how to set up different scenarios and then check that `Runtime::availableProcessors()` reports what would be expected. Otherwise we have a ton of code here that is effectively untested.

I have made some suggestions for uniform logging statements so that we could use UL together with the test scenarios to check whether things are working as expected.

Thanks.

dholmes-ora reviewed (commented):
Initialization update looks good. Thanks.

swesonga commented:
> Sorry for the delay in getting back to this. There is a lot of API surface that I am not familiar with that I am trying to get my head around. The overall code structure and approach is looking very good - though I have one issue with the initialization process.
> 
> One thing that is missing though is a way to effectively test this given all the possible conditions that might be present:
> 
> * a job object exists, or not
> * processor groups exist, or not
> * affinity has been set, or not
> 
> I recognise it may be very difficult to create either a regression test or a gtest for this, but I would like to see some instructions on how to set up different scenarios and then check that `Runtime::availableProcessors()` reports what would be expected. Otherwise we have a ton of code here that is effectively untested.
> 
> I have made some suggestions for uniform logging statements so that we could use UL together with the test scenarios to check whether things are working as expected.
> 
> Thanks.

I have updated the PR with logging code and added a "Test Scenarios" section to the PR description outlining these scenarios. Please let me know if there is a better place to save this information. As far as I can tell, there is no standard Windows tool for creating job objects though.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -860,6 +860,8 @@ bool os::has_allocatable_memory_limit(size_t* limit) {
 }
 
 int os::active_processor_count() {
+  win32::log_general_system_processor_debug_info();
```
Please move this to the init function as we don't want to have to check logging on each call. We don't need a separate function in that case. Thanks

dholmes-ora commented:
> I have updated the PR with logging code and added a "Test Scenarios" section to the PR description outlining these scenarios.

Thank you for that. I'm still wrestling with how we can test this.

dholmes-ora reviewed (commented):
Updates look good thanks.

dholmes-ora commented:
I wonder if @mo-beck might be able to get someone to review this who has more experience with this aspect of Windows? Thanks in advance.

dholmes-ora reviewed (commented):
Version check update looks good.

swesonga commented:
@dholmes-ora I have received some feedback from folks on the Windows team. There will be a few more changes coming next week.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -4211,7 +4168,13 @@ void os::win32::initialize_system_info() {
   bool schedules_all_processor_groups = win32::is_windows_11_or_greater() || win32::is_windows_server_2022_or_greater();
   if (schedules_all_processor_groups) {
-    processors = system_logical_processor_count();
+    processors = GetActiveProcessorCount(ALL_PROCESSOR_GROUPS);
```
This doesn't seem right. Don't we want to know the number of processors available to the JVM not the number in the entire "system"?

dholmes-ora commented on test/hotspot/jtreg/runtime/6942632/exeGetProcessorInfo.c:
The compilation of the native part of this test should only happen on windows. You need to add appropriate instructions to build it, and appropriate instructions to exclude it on non-windows in make/test/JtregNativeHotspot.gmk. Then you don't need the win64 ifdef guards.

dholmes-ora reviewed (changes requested):
Thanks for adding the test but it needs to be adapted to how we write/execute the tests today. See comments below.

dholmes-ora reviewed (commented):
Test changes look good. Thanks. I will run this through our CI.

swesonga commented:
> Test changes look good. Thanks. I will run this through our CI.

@dholmes-ora, any word on whether this change passed the CI tests?

dholmes-ora commented:
@swesonga  apologies I only intended to report back if there were any problems. Things tested fine.

swesonga commented:
@dholmes-ora, @jdksjolen, are there any other changes needed to get this PR approved?

dholmes-ora reviewed (approved):
Nothing further from me. Thanks

swesonga commented:
/integrate

jdksjolen reviewed (approved):
Hi,

This looks good to me, thanks.

dholmes-ora commented:
/sponsor
